internal/uvm/vnuma.go

```diff
@@ -0,0 +1,166 @@
+//go:build windows
+
+package uvm
+
+import (
+	"fmt"
+
+	hcsschema "github.com/Microsoft/hcsshim/internal/hcs/schema2"
+	"github.com/Microsoft/hcsshim/osversion"
+)
+
+// prepareVNumaTopology creates vNUMA settings for implicit (platform) or explicit (user-defined) topology.
+//
+// For implicit topology we look for `MaxProcessorsPerNumaNode` and `MaxSizePerNode` create options. Setting them
+// in HCS doc, will trigger platform to create vNUMA topology based on the given values. Based on experiments, the
+// platform will create an evenly distributed topology based on requested memory and processor count for the HCS VM.
+//
+// For explicit topology we look for `NumaMappedPhysicalNodes`, `NumaProcessorCounts` and `NumaMemoryBlocksCounts` create
+// options. The above options are number slices, where a value at index `i` in each slice represents the corresponding
+// value for the `i`th vNUMA node.
+// Limitations:
+// - only hcsschema.MemoryBackingType_PHYSICAL is supported
+// - `PhysicalNumaNodes` values at index `i` will be mapped to virtual node number `i`
+// - client is responsible for setting wildcard physical node numbers
+func prepareVNumaTopology(opts *Options) (*hcsschema.Numa, *hcsschema.NumaProcessors, error) {
+	if opts.MaxProcessorsPerNumaNode == 0 && len(opts.NumaMappedPhysicalNodes) == 0 {
+		// vNUMA settings are missing, return empty topology
+		return nil, nil, nil
+	}
+
+	var preferredNumaNodes []int64
+	for _, pn := range opts.PreferredPhysicalNumaNodes {
+		preferredNumaNodes = append(preferredNumaNodes, int64(pn))
+	}
+
+	build := osversion.Get().Build
+
+	// Implicit vNUMA topology.
+	if opts.MaxProcessorsPerNumaNode > 0 {
+		if build < osversion.V25H1Server {
+			return nil, nil, fmt.Errorf("implicit vNUMA topology is not supported on %d version of Windows", build)
+		}
+		if opts.MaxSizePerNode == 0 {
+			return nil, nil, fmt.Errorf("max size per node must be set when max processors per numa node is set")
+		}
+		numaProcessors := &hcsschema.NumaProcessors{
+			CountPerNode: hcsschema.Range{
+				Max: opts.MaxProcessorsPerNumaNode,
+			},
+		}
+		numa := &hcsschema.Numa{
+			MaxSizePerNode:         opts.MaxSizePerNode,
+			PreferredPhysicalNodes: preferredNumaNodes,
+		}
+		return numa, numaProcessors, nil
+	}
+
+	// Explicit vNUMA topology.
+
+	numaNodeCount := len(opts.NumaMappedPhysicalNodes)
+	if numaNodeCount != len(opts.NumaProcessorCounts) || numaNodeCount != len(opts.NumaMemoryBlocksCounts) {
+		return nil, nil, fmt.Errorf("mismatch in number of physical numa nodes and the corresponding processor and memory blocks count")
+	}
+
+	if build < osversion.V25H1Server {
```

Comment:
can't we have this check at the top of the function, since we hit this check in both implicit and explicit cases?

Reply:
Done.